### ISSUE TITLE
ITVplayer plugin issues

### ISSUE THREAD

**Issue by Vangelis66** [chrippa/livestreamer#1233]
Hi - many thanks for this app!

OS: WinVistaSP2 x86
Livestreamer: zipped version of latest nightly, i.e. 
livestreamer-v1.12.2-120-gab80dbd-20160202222614Z
rtmpdump: 2.4-20151223-gfa8646d-OpenSSL_1.0.2f-x86-static
(non-UK physical location)

Trying to dump the ITV live streams.
1. First of all, ITVplayer was rebranded 
   last November 2015 to ITV Hub.

The pages containing the live streams are now: 

http://www.itv.com/hub/itv
http://www.itv.com/hub/itv2
http://www.itv.com/hub/itvbe
http://www.itv.com/hub/itv3
http://www.itv.com/hub/itv4
http://www.itv.com/hub/citv

Of course, these new URLs are not picked up 
by existing itvplayer plugin (itvplayer.py)

 Second change is that registration is now 
mandatory to access streams in a browser...

 One other notable change is that itv (ITV1, 
main TV channel), is now being streamed via 
AdobeHDS (with no Live-Rewind) method. 
The generation of the playlist/manifest requires 
a UK IP, but the hostname used for both 
the hds manifest and hds fragments is not 
geo-blocked ("itv1livehds-lh.itv.com"). 
I can presently fetch this live stream if I first find 
the f4m manifest and then feed it to livestreamer; 
hopefully, when patched, livestreamer could just parse 
`http://www.itv.com/hub/itv` 

 I am using a UK proxy just for playlist generation, 
so I would hope that, e.g. 
`livestreamer --http-proxy "proxyhost:proxyport" -o "itv1-1800.flv" "http://www.itv.com/hub/itv" best` could use the proxy just for that 
and then fetch the stream itself with a DIRECT connection 
(although I read and understood that --http-proxy 
is used for all HTTP requests :-( ; possibly what I ask 
could be fixed inside the plugin, or an additional 
--http-proxy-partial switch could be created for cases 
like these; [get-iplayer](https://github.com/get-iplayer/get_iplayer) does have a similar switch...).

The remaining channels, as far as I can tell, 
still employ RTMPE streaming method,so a 
workaround with the existing plugin is 
to use the "old-style" URL templates: 

https://itv.com/itvplayer/itv2
https://itv.com/itvplayer/itvbe
https://itv.com/itvplayer/itv3
https://itv.com/itvplayer/itv4
https://itv.com/itvplayer/citv

All the RTMP hostnames are not geo-blocked, so e.g. 

`livestreamer -l debug --http-proxy "proxyhost:proxyport" -o "itvbe-1800.flv" "https://itv.com/itvplayer/itvbe" best` works from my overseas location: 

```
[cli][info] Found matching plugin itvplayer for URL https://itv.com/itvplayer/it
vbe
[cli][info] Available streams: 600k (worst), 800k, 1200k, 1500k, 1800k (best)
[cli][info] Opening stream: 1800k (rtmp)
[cli][debug] Pre-buffering 8192 bytes
[cli][debug] Checking file output
[cli][debug] Writing stream to output
[download][itvbe-1800.flv] Written 7.8 MB (26s @ 212.0 KB/s)                  
```

As explained already, using `https://itv.com/itvplayer/itv[1]` 
just produces an error currently:
`error: Unable to open URL: https://itv.com/itvplayer/itv (404 Client Error: Not Found)`

In anticipation of fixes/patches, many thanks again 
for all your efforts...

**Comment by tboss** [comment]
Hi,

Just made a new pull request for this issue, where the URL matching is fixed and HDS stream compatibility has been added. Good, detailed issue report by the way.

As you mention, the geo-restriction applies only (at least currently) to the initial playlist/response data. I also had to use a proxy via --http-proxy switch (as I don't have access to a UK IP), and as you note it would be nice to be able to disable the proxy when it is no longer needed. As a crude way of achieving that, I put a commented region in the code after the initial response data has been fetched. Try it out if you want, by uncommenting those lines. I does work, but a command-line flag solution is probably better than simply removing the proxies.

/tboss

**Comment by Vangelis66** [comment]
@tboss wrote:

> Just made a new pull request for this issue

Hello tboss, much obliged really!

> Good, detailed issue report by the way.

... Many thanks for your praise...

> where the URL matching is fixed

Sadly, I am not able to write Python code, 
in some cases I can just... pretend to understand 
parts of already written one.... 

Since there's no longer an "ITVPlayer" site per se, 
I thought it was more appropriate to rename the 
plugin to "ITVhub"; so I fetched your patched version from:

[itvplayer.py](https://github.com/tboss/livestreamer/raw/patch/plugin_itvplayer/src/livestreamer/plugins/itvplayer.py)

renamed it and then applied following 
patch: [itvhub.patch](https://github.com/chrippa/livestreamer/files/172221/itvhub.patch.txt) (I hope you don't mind). 
Then "itvhub.py" was placed inside
`livestreamer-v1.12.2-120-gab80dbd\livestreamer-1.12.2-py2.7.egg\livestreamer\plugins`
where it works as expected!

> and HDS stream compatibility has been added

Well, in the 18 days that have passed since 
my original post, they seem to have reverted 
ITV1 (itv) streaming again back to RTMPE, 
which is indeed quite odd, having used AdobeHDS 
for that one for at least five months; 
so unable now to test your HDS implementation, 
many thanks regardless!

`livestreamer -l debug --http-proxy "host:port" "http://www.itv.com/hub/itv" best -c` =>

 `rtmpdump.exe --live --playpath="itv1livefms_1800@s172940?auth=eaEcZagcwdwbfcUbJcacPa3bub7dkcraJavcccAcCdWdgbmdHaWa_cvcGc_bQazcfao-bw5G0P-frG-bCKgEDpkA0Gatcup&aifp=v001&slist=itv1livefms_1800@s172940" --rtmp="rtmpe://cp274438.live.edgefcs.net/live" --timeout="10" --app="live" --swfVfy="http://www.itv.com/mediaplayer/ITVMediaPlayer.swf"`

Many kind regards!

**Comment by tboss** [comment]
> Since there's no longer an "ITVPlayer" site per se, 
> I thought it was more appropriate to rename the 
> plugin to "ITVhub"

You're right, it probably should be renamed as even the wikipedia article mentions this: https://en.wikipedia.org/wiki/ITV_Hub . It might even be best to rename it ITV (e.g. itv.py for the plugin file) if it can handle all of ITV's (free) streaming content (?). Although, worth noting, the plugin will of course work regardless of the actual name of the plugin (actual file or class name) -- basically if the `can_handle_url()` method returns `True` the plugin can handle the URL, otherwise not. I'll probably upload a patch for the name change though, or at least add a comment to the pull request.

> Well, in the 18 days that have passed since 
> my original post, they seem to have reverted 
> ITV1 (itv) streaming again back to RTMPE

That seems to be the case, yes. I discovered that while testing the patched version last night. Though as I made some of the code changes, including the HDS parsing, soon after your post (when they still used AdobeHDS in some fashion), I can confirm the HDS manifest was properly parsed and streams popped out. Maybe they'll reintroduce it, and in that case the code should handle it properly.

Sidenote for Linux: A strange thing I noticed was that under my Manjaro Linux boot, in which I wrote the patch, the stream quality (e.g. 600k, 800k, 1200k, etc.) would not indicate best/worst, thus (seemingly) failing at evaluating the qualities, and also missing the "k" after the number. Here, on my Windows boot the same code works as intended, with "k" and indicating best/worst. So I'm not sure exactly what is going on with that.
